File: typescript/src/client/SaeClientOptions.ts
import { AxiosRequestConfig } from 'axios';

/**
 * Nivel de log para eventos del cliente SAE.
 */
export enum SaeLogLevel {
    Debug = 'debug',
    Info = 'info',
    Warning = 'warning',
    Error = 'error'
}

/**
 * Evento de log para el cliente SAE.
 */
export interface SaeClientLogEvent {
    timestamp: Date;
    level: SaeLogLevel;
    message: string;
    operation?: string;
    retryAttempt?: number;
    error?: Error;
    duration?: number; // en milisegundos
}

/**
 * Opciones de configuración avanzadas para el cliente SAE TypeScript.
 */
export interface SaeClientOptions {
    /**
     * Timeout para las peticiones HTTP en milisegundos.
     * Por defecto: 100000 (100 segundos).
     */
    timeout?: number;

    /**
     * Número máximo de reintentos en caso de fallos temporales (5xx, timeout, etc.).
     * Por defecto: 3 reintentos.
     */
    maxRetries?: number;

    /**
     * Delay inicial entre reintentos en milisegundos (backoff exponencial).
     * Por defecto: 1000 (1 segundo).
     */
    retryDelay?: number;

    /**
     * Factor de multiplicación para el backoff exponencial.
     * Por defecto: 2 (1s, 2s, 4s, etc.)
     */
    backoffMultiplier?: number;

    /**
     * Indica si se debe ignorar la validación de certificados SSL.
     * ⚠️ SOLO USAR EN DESARROLLO con certificados autofirmados.
     * Por defecto: false.
     * 
     * En Node.js, esto configura rejectUnauthorized en el HTTPS agent.
     * En navegador, esto no tiene efecto (la seguridad es manejada por el navegador).
     */
    skipCertificateValidation?: boolean;

    /**
     * Certificado de cliente para autenticación mTLS (solo Node.js).
     * Formato: { cert: string, key: string, ca?: string }
     */
    clientCertificate?: {
        cert: string;
        key: string;
        ca?: string;
    };

    /**
     * Certificados CA raíz de confianza personalizados (solo Node.js).
     * Array de strings PEM o buffers.
     */
    customRootCertificates?: string[] | Buffer[];

    /**
     * Agente HTTPS personalizado (solo Node.js).
     * Si se proporciona, se usa en lugar de las opciones de certificado.
     */
    httpsAgent?: any;

    /**
     * Número máximo de redirecciones permitidas.
     * Por defecto: 5.
     */
    maxRedirects?: number;

    /**
     * Indica si se deben seguir las redirecciones automáticamente.
     * Por defecto: true.
     */
    followRedirects?: boolean;

    /**
     * Callback para logging de eventos de red (debug, errores, reintentos).
     */
    onLog?: (event: SaeClientLogEvent) => void;

    /**
     * Configuración adicional de Axios.
     * Estas opciones se fusionan con la configuración interna.
     */
    axiosConfig?: Partial<AxiosRequestConfig>;
}

/**
 * Valores predeterminados para producción.
 */
export const ProductionDefaults: Required<Pick<SaeClientOptions, 
    'timeout' | 'maxRetries' | 'retryDelay' | 'backoffMultiplier' | 
    'skipCertificateValidation' | 'maxRedirects' | 'followRedirects'>> = {
    timeout: 100000,
    maxRetries: 3,
    retryDelay: 1000,
    backoffMultiplier: 2,
    skipCertificateValidation: false,
    maxRedirects: 5,
    followRedirects: true
};

/**
 * Valores predeterminados para desarrollo.
 * ⚠️ No usar en producción.
 */
export const DevelopmentDefaults: Required<Pick<SaeClientOptions,
    'timeout' | 'maxRetries' | 'retryDelay' | 'backoffMultiplier' | 
    'skipCertificateValidation' | 'maxRedirects' | 'followRedirects'>> = {
    timeout: 30000,
    maxRetries: 1,
    retryDelay: 500,
    backoffMultiplier: 2,
    skipCertificateValidation: true, // ⚠️ Solo para desarrollo
    maxRedirects: 5,
    followRedirects: true
};

/**
 * Excepción específica para errores del cliente SAE.
 */
export class SaeClientError extends Error {
    /**
     * Número de intentos realizados antes de fallar.
     */
    readonly retryCount: number;

    /**
     * Código de error HTTP si aplica.
     */
    readonly statusCode?: number;

    constructor(message: string, retryCount: number = 0, statusCode?: number) {
        super(message);
        this.name = 'SaeClientError';
        this.retryCount = retryCount;
        this.statusCode = statusCode;
    }
}

/**
 * Determina si un error de HTTP es retentable.
 */
export function isRetryableError(error: any): boolean {
    // Errores de red/timeout
    if (error.code === 'ECONNRESET' || 
        error.code === 'ETIMEDOUT' || 
        error.code === 'ECONNREFUSED' ||
        error.code === 'ENOTFOUND' ||
        error.code === 'EAI_AGAIN' ||
        error.code === 'ENETUNREACH') {
        return true;
    }

    // Errores de timeout de axios
    if (error.code === 'ECONNABORTED' && error.message?.includes('timeout')) {
        return true;
    }

    // Errores 5xx
    if (error.response && error.response.status >= 500 && error.response.status < 600) {
        return true;
    }

    // 429 Too Many Requests
    if (error.response && error.response.status === 429) {
        return true;
    }

    // 408 Request Timeout
    if (error.response && error.response.status === 408) {
        return true;
    }

    // 502 Bad Gateway, 503 Service Unavailable, 504 Gateway Timeout
    if (error.response && (error.response.status === 502 || 
                            error.response.status === 503 || 
                            error.response.status === 504)) {
        return true;
    }

    return false;
}

/**
 * Calcula el delay para el siguiente reintento usando backoff exponencial con jitter.
 */
export function calculateRetryDelay(attempt: number, baseDelay: number, multiplier: number): number {
    // Backoff exponencial: delay * (multiplier ^ attempt)
    const exponentialDelay = baseDelay * Math.pow(multiplier, attempt);
    
    // Agregar jitter aleatorio (±25%) para evitar thundering herd
    const jitter = exponentialDelay * 0.25 * (Math.random() * 2 - 1);
    const finalDelay = exponentialDelay + jitter;
    
    return Math.max(finalDelay, 100); // Mínimo 100ms
}

File: csharp/Client/SaeClientOptions.cs
using System.Net;
using System.Net.Security;
using System.Security.Authentication;
using System.Security.Cryptography.X509Certificates;

namespace SAE.Sdk.Client;

/// <summary>
/// Opciones de configuración avanzadas para el cliente SAE.
/// </summary>
public class SaeClientOptions
{
    /// <summary>
    /// Timeout para las peticiones HTTP. Por defecto: 100 segundos.
    /// </summary>
    public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(100);

    /// <summary>
    /// Número máximo de reintentos en caso de fallos temporales (5xx, timeout, etc.).
    /// Por defecto: 3 reintentos.
    /// </summary>
    public int MaxRetries { get; set; } = 3;

    /// <summary>
    /// Delay inicial entre reintentos (backoff exponencial).
    /// Por defecto: 1 segundo.
    /// </summary>
    public TimeSpan RetryDelay { get; set; } = TimeSpan.FromSeconds(1);

    /// <summary>
    /// Factor de multiplicación para el backoff exponencial.
    /// Por defecto: 2 (1s, 2s, 4s, etc.)
    /// </summary>
    public double BackoffMultiplier { get; set; } = 2.0;

    /// <summary>
    /// Tiempo máximo de espera total para todos los reintentos.
    /// Por defecto: 30 segundos.
    /// </summary>
    public TimeSpan MaxRetryDuration { get; set; } = TimeSpan.FromSeconds(30);

    /// <summary>
    /// Indica si se debe ignorar la validación de certificados SSL.
    /// ⚠️ SOLO USAR EN DESARROLLO con certificados autofirmados.
    /// Por defecto: false.
    /// </summary>
    public bool SkipCertificateValidation { get; set; } = false;

    /// <summary>
    /// Callback personalizado para validar certificados SSL.
    /// Permite validación personalizada de certificados (thumbprints, CAs específicos, etc.)
    /// </summary>
    public Func<X509Certificate2, X509Chain, SslPolicyErrors, bool>? CustomCertificateValidation { get; set; }

    /// <summary>
    /// Certificado de cliente para autenticación mTLS (mutual TLS).
    /// Opcional: se usa cuando el servidor requiere autenticación por certificado.
    /// </summary>
    public X509Certificate2? ClientCertificate { get; set; }

    /// <summary>
    /// Colección de certificos de CA raíz de confianza personalizados.
    /// Útil para validar certificados emitidos por CAs internas/empresariales.
    /// </summary>
    public X509Certificate2Collection? CustomRootCertificates { get; set; }

    /// <summary>
    /// Indica si se debe habilitar el Keep-Alive para conexiones persistentes.
    /// Por defecto: true.
    /// </summary>
    public bool EnableKeepAlive { get; set; } = true;

    /// <summary>
    /// Tiempo de vida máximo de una conexión persistente.
    /// Por defecto: 2 minutos.
    /// </summary>
    public TimeSpan PooledConnectionLifetime { get; set; } = TimeSpan.FromMinutes(2);

    /// <summary>
    /// Número máximo de conexiones por servidor.
    /// Por defecto: int.MaxValue (sin límite práctico).
    /// </summary>
    public int MaxConnectionsPerServer { get; set; } = int.MaxValue;

    /// <summary>
    /// Indica si se debe seguir redirecciones automáticamente.
    /// Por defecto: true.
    /// </summary>
    public bool AllowAutoRedirect { get; set; } = true;

    /// <summary>
    /// Número máximo de redirecciones permitidas.
    /// Por defecto: 50.
    /// </summary>
    public int MaxAutomaticRedirections { get; set; } = 50;

    /// <summary>
    /// Versión de TLS mínima permitida.
    /// Por defecto: TLS 1.2
    /// </summary>
    public SslProtocols MinSslProtocol { get; set; } = SslProtocols.Tls12;

    /// <summary>
    /// Callback para logging de eventos de red (debug, errores, reintentos).
    /// </summary>
    public Action<SaeClientLogEvent>? LogCallback { get; set; }

    /// <summary>
    /// Crea opciones predeterminadas seguras para producción.
    /// </summary>
    public static SaeClientOptions ProductionDefaults() => new()
    {
        Timeout = TimeSpan.FromSeconds(100),
        MaxRetries = 3,
        SkipCertificateValidation = false,
        MinSslProtocol = SslProtocols.Tls12
    };

    /// <summary>
    /// Crea opciones para desarrollo local con certificados autofirmados.
    /// ⚠️ No usar en producción.
    /// </summary>
    public static SaeClientOptions DevelopmentDefaults() => new()
    {
        Timeout = TimeSpan.FromSeconds(30),
        MaxRetries = 1,
        SkipCertificateValidation = true,
        MinSslProtocol = SslProtocols.Tls12
    };
}

/// <summary>
/// Evento de log para el cliente SAE.
/// </summary>
public class SaeClientLogEvent
{
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
    public SaeLogLevel Level { get; set; }
    public string Message { get; set; } = "";
    public string? Operation { get; set; }
    public int? RetryAttempt { get; set; }
    public Exception? Exception { get; set; }
    public TimeSpan? Duration { get; set; }
}

public enum SaeLogLevel
{
    Debug,
    Info,
    Warning,
    Error
}

/// <summary>
/// Excepción específica para errores del cliente SAE.
/// </summary>
public class SaeClientException : Exception
{
    /// <summary>
    /// Número de intentos realizados antes de fallar.
    /// </summary>
    public int RetryCount { get; }

    public SaeClientException(string message) : base(message)
    {
        RetryCount = 0;
    }

    public SaeClientException(string message, Exception? innerException) : base(message, innerException)
    {
        RetryCount = 0;
    }

    public SaeClientException(string message, Exception? innerException, int retryCount) : base(message, innerException)
    {
        RetryCount = retryCount;
    }
}
